core/drivers/nxp_ls_i2c.c

```diff
@@ -0,0 +1,420 @@
+	if (status != TEE_SUCCESS)
+		return status;
+
+	// Write Data
+	for (index = 0; index < operation->length_in_bytes; index++) {
+		io_write8((vaddr_t)&regs->ibdr, operation->buffer[index]);
+		status = i2c_transfer_complete(regs, I2C_BUS_TEST_RX_ACK);
+		if (status != TEE_SUCCESS)
+			return status;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result i2c_stop(struct i2c_regs *regs)
+{
+	TEE_Result status = TEE_ERROR_GENERIC;
+	uint8_t reg = 0;
+
+	status = TEE_SUCCESS;
+	reg = io_read8((vaddr_t)&regs->ibsr);
+	if (reg & I2C_IBSR_IBB) {
+		// Generate Stop Signal
+		i2c_io_and8((vaddr_t)&regs->ibcr,
+			    ~(I2C_IBCR_MSSL | I2C_IBCR_TXRX));
+		status = i2c_bus_test_bus_busy(regs, I2C_BUS_TEST_IDLE);
+		if (status != TEE_SUCCESS)
+			return status;
+	}
+
+	// Disable I2c Controller
+	i2c_io_or8((vaddr_t)&regs->ibcr, I2C_IBCR_MDIS);
+
+	return status;
+}
+
+static TEE_Result i2c_start(struct i2c_regs *regs)
+{
+	TEE_Result status = TEE_ERROR_GENERIC;
+
+	i2c_io_or8((vaddr_t)&regs->ibsr, (I2C_IBSR_IBAL | I2C_IBSR_IBIF));
+	i2c_io_and8((vaddr_t)&regs->ibcr, (uint8_t)(~I2C_IBCR_MDIS));
+
+	//Wait controller to be stable
+	mdelay(1);
+
+	// Generate Start Signal
+	i2c_io_or8((vaddr_t)&regs->ibcr, I2C_IBCR_MSSL);
+	status = i2c_bus_test_bus_busy(regs, I2C_BUS_TEST_BUSY);
+	if (status != TEE_SUCCESS)
+		return status;
+
+	// Select Transmit Mode. set No ACK = 1
+	i2c_io_or8((vaddr_t)&regs->ibcr, (I2C_IBCR_TXRX | I2C_IBCR_NOACK));
+
+	return status;
+}
+
+/*
+ * Transfer data to/from I2c slave device
+ * Base           Base Address of I2c controller's registers
+ * SlaveAddress   Slave Address from which data is to be read
+ * RequestPacket  Pointer to an EFI_I2C_REQUEST_PACKET structure
+ * describing the I2C transaction
+ */
+TEE_Result i2c_bus_xfer(vaddr_t base, uint32_t slave_address,
+			struct i2c_reg_request *request_packet)
+{
+	uint32_t index = 0;
+	struct i2c_regs *regs = (struct i2c_regs *)base;
+	struct i2c_operation *operation;
+	TEE_Result status = TEE_ERROR_GENERIC;
+	bool is_last_operation = FALSE;
+
+	status = i2c_bus_test_bus_busy(regs, I2C_BUS_TEST_IDLE);
+	if (status != TEE_SUCCESS)
+		goto error_exit;
+
+	status = i2c_start(regs);
+	if (status != TEE_SUCCESS)
+		goto error_exit;
+
+	for (index = 0, operation = request_packet->operation;
+	     index < request_packet->operation_count; index++, operation++) {
+		if (index == (request_packet->operation_count - 1))
+			is_last_operation = TRUE;
+
+		// Send repeat start after first transmit/receive
+		if (index) {
+			i2c_io_or8((vaddr_t)&regs->ibcr, I2C_IBCR_RSTA);
+			status = i2c_bus_test_bus_busy(regs, I2C_BUS_TEST_BUSY);
+			if (status != TEE_SUCCESS)
+				goto error_exit;
+		}
+
+		// Read/write data
+		if (operation->flags & I2C_FLAG_READ)
+			status = i2c_read(regs, slave_address, operation,
+					  is_last_operation);
+		else
+			status = i2c_write(regs, slave_address, operation);
+		if (status != TEE_SUCCESS)
+			goto error_exit;
+	}
+
+error_exit:
```

Comment:
This isn't accurate, we end up here on success too. A better name would be `out:`, that's what we normally use.